vms/platformvm/block/executor/warp_verifier.go

```diff
@@ -0,0 +1,95 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package executor
+
+import (
+	"context"
+
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/vms/platformvm/block"
+	"github.com/ava-labs/avalanchego/vms/platformvm/txs"
+	"github.com/ava-labs/avalanchego/vms/platformvm/txs/executor"
+)
+
+var _ block.Visitor = (*warpVerifier)(nil)
+
+// VerifyWarpMessages verifies all warp messages in the block. If any of the
+// warp messages are invalid, an error is returned.
+func VerifyWarpMessages(
+	ctx context.Context,
+	networkID uint32,
+	validatorState validators.State,
+	pChainHeight uint64,
+	b block.Block,
+) error {
+	return b.Visit(&warpVerifier{
+		ctx:            ctx,
+		networkID:      networkID,
+		validatorState: validatorState,
+		pChainHeight:   pChainHeight,
+	})
+}
+
+type warpVerifier struct {
+	ctx            context.Context
+	networkID      uint32
+	validatorState validators.State
+	pChainHeight   uint64
+}
+
+func (*warpVerifier) BanffAbortBlock(*block.BanffAbortBlock) error {
+	return nil
+}
+
+func (*warpVerifier) BanffCommitBlock(*block.BanffCommitBlock) error {
+	return nil
+}
+
+func (w *warpVerifier) BanffProposalBlock(blk *block.BanffProposalBlock) error {
+	for _, tx := range blk.Transactions {
+		if err := w.verify(tx); err != nil {
+			return err
+		}
+	}
+	return w.ApricotProposalBlock(&blk.ApricotProposalBlock)
```

Comment:
Why do we need warpVerifier in banff blocks? Also aren't we repeating the same verification with w.verify in BanffProposalBlock + and one in the ApricotProposalBlock?

Reply:
> Why do we need warpVerifier in banff blocks?

These are called `Banff` blocks because they were introduced in `Banff`. They are the latest block type (so the P-chain will continue to issue `Banff` blocks after `Etna` is activated).

> Also aren't we repeating the same verification with w.verify in BanffProposalBlock + and one in the ApricotProposalBlock?

`w.verify` takes in a transaction, for a single block, we call `w.verify` for every transaction in the block.